src/compile/legend/encode.ts

```diff
@@ -82,6 +83,20 @@ export function labels(fieldDef: FieldDef<string>, labelsSpec: any, model: UnitM
         signal: timeFormatExpression('datum.value', fieldDef.timeUnit, legend.format, config.legend.shortTimeLabels, config.timeFormat, isUTCScale)
       }
     }, labelsSpec || {});
+  } else if ((fieldDef.type === NOMINAL || fieldDef.type === ORDINAL) && legend.format) {
```

Comment:
This chunk of code seems identical to axis and very similar to `formatSignalRef`.  

I'd say try to extract a common method for both axis and legend  (name it `guideLabels` text, which return `{signal: ...}`.   And if possible, see if there is a way to reuse `formatSignalRef`.  Note that the latter is trickier since we only set `labels.text` when `format` property of axis/legend does not natively work. So I'm fine if you only do the former first (for now).